File: frontend/src/services/jump-types.ts
import { api } from '@/lib/api';
import { JumpType, UserRole } from '@/types';

export interface JumpTypesListParams {
  skip?: number;
  limit?: number;
  allowed_role?: UserRole;
  is_available?: boolean;
  search?: string;
}

export interface CreateJumpTypeData {
  name: string;
  short_name: string;
  description?: string;
  exit_altitude?: number;
  price?: number;
  is_available?: boolean;
  allowed_roles?: UserRole[];
  additional_staff?: Array<{
    staff_required_role: UserRole;
    staff_default_jump_type_id?: number;
  }>;
}

export interface UpdateJumpTypeData {
  name?: string;
  short_name?: string;
  description?: string;
  exit_altitude?: number;
  price?: number;
  is_available?: boolean;
  allowed_roles?: UserRole[];
  additional_staff?: Array<{
    staff_required_role: UserRole;
    staff_default_jump_type_id?: number;
  }>;
}

export const jumpTypesService = {
  // Get all jump types with optional filters
  getJumpTypes: async (params: JumpTypesListParams = {}): Promise<JumpType[]> => {
    const response = await api.get('/jump-types/', { params });
    return response.data;
  },

  // Get single jump type by ID
  getJumpType: async (id: number): Promise<JumpType> => {
    const response = await api.get(`/jump-types/${id}`);
    return response.data;
  },

  // Create new jump type (admin only)
  createJumpType: async (data: CreateJumpTypeData): Promise<JumpType> => {
    const response = await api.post('/jump-types/', data);
    return response.data;
  },

  // Update jump type (admin only)
  updateJumpType: async (id: number, data: UpdateJumpTypeData): Promise<JumpType> => {
    const response = await api.put(`/jump-types/${id}`, data);
    return response.data;
  },

  // Delete jump type (admin only)
  deleteJumpType: async (id: number): Promise<void> => {
    await api.delete(`/jump-types/${id}`);
  },
};

File: backend/app/api/v1/jump_types.py
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from app.api.deps import get_current_user, get_admin_user
from app.core.database import get_db
from app.crud.jump_types import jump_type as jump_type_crud
from app.schemas.jump_types import JumpTypeResponse, JumpTypeUpdate, JumpTypeCreate
from app.models.users import User
from app.models.enums import UserRole

router = APIRouter()


#=========================#
#                         #
#   Jump Type Endpoints   #
#                         #
#=========================#

@router.get("/", response_model=List[JumpTypeResponse])
def list_jump_types(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    allowed_role: UserRole = Query(None),
    is_available: bool = Query(None),
    search: str = Query(None, min_length=2),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """List all jump types with optional filters"""
    filters = {}
    if search:
        filters['search'] = search
    if allowed_role:
        filters['allowed_role'] = allowed_role
    if is_available is not None:
        filters['is_available'] = is_available
    
    return jump_type_crud.get_jump_types(db, filters=filters, skip=skip, limit=limit)


@router.get("/{jump_type_id}", response_model=JumpTypeResponse)
def read_jump_type(
    jump_type_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user)
):
    """Get jump type by ID"""
    jump_type = jump_type_crud.get(db, id=jump_type_id)
    if not jump_type:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Jump type not found"
        )
    return jump_type


#=========================#
#                         #
#   Admin Endpoints       #
#                         #
#=========================#

@router.post("/", response_model=JumpTypeResponse)
def create_jump_type(
    jump_type_create: JumpTypeCreate,
    db: Session = Depends(get_db),
    admin_user: User = Depends(get_admin_user)
):
    """Create a new jump type (admin only)"""
    return jump_type_crud.create(db, obj_in=jump_type_create, created_by=admin_user.id)


@router.put("/{jump_type_id}", response_model=JumpTypeResponse)
def update_jump_type(
    jump_type_id: int,
    jump_type_update: JumpTypeUpdate,
    db: Session = Depends(get_db),
    admin_user: User = Depends(get_admin_user)
):
    """Update jump type including allowed roles and additional staff (admin only)"""
    jump_type = jump_type_crud.get(db, id=jump_type_id)
    if not jump_type:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Jump type not found"
        )
    
    return jump_type_crud.update(
        db,
        db_obj=jump_type,
        obj_in=jump_type_update,
        updated_by=admin_user.id
    )


@router.delete("/{jump_type_id}")
def delete_jump_type(
    jump_type_id: int,
    db: Session = Depends(get_db),
    admin_user: User = Depends(get_admin_user)
):
    """Delete jump type (admin only)"""
    jump_type = jump_type_crud.get(db, id=jump_type_id)
    if not jump_type:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Jump type not found"
        )
    
    jump_type_crud.remove(db, id=jump_type_id, deleted_by=admin_user.id)
    return {"message": "Jump type deleted successfully"}
